lighthouse-core/gather/gatherers/forms.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ *  @param {HTMLFormElement} formElement
+ *  @return { {inputs: LH.Artifacts['FormInputs'][], labels: LH.Artifacts['FormLabels'][]}}
+ */
+/* istanbul ignore next */
+function getChildrenElements(formElement) {
+  /** @type {LH.Artifacts['FormInputs'][]} */
+  const inputEls = [];
+  /** @type {LH.Artifacts['FormLabels'][]} */
+  const labels = [];
+  const childrenArray = /** @type {HTMLElement[]} */ ([...formElement.childNodes]);
```

Comment:
two things relating to how we process all these elements.

1. this childNodes thing is a little brittle. folks have all sorts of `<div>`s, `<fieldset>`s in their forms. so we'd at least need to `form.querySelectorAll('*)`. but I dont think that makes sense because.... 
2. i agree with @connorjclark's flow in https://github.com/GoogleChrome/lighthouse/pull/11062#discussion_r456714689 and was expecting a slightly different thing...


down in your `for (const child of formChildren) {` in `collectFormElements()`... it seems like we already have all the elements we care about. we don't need collect elements all over again in `getChildrenElements` (and dont need to repeat these instanceof checks).

i was thinking something along these lines:

```js
  for (const child of formChildren) {

    const hasDomForm = child.form;
	
    // Initalize an form object if there's not already one
    if (hasDomForm && !forms.has(child.form)) {
      const newformObj = {attributes: {
          id: child.form.id,
          name: child.form.name,
          autocomplete: child.form.autocomplete,
        },
        inputs: [],
        labels: [],
      };
      forms.set(child.form, newformObj);
    }

    const formObj = hasDomForm ? forms.get(child.form) : formlessObj;

	// Now add child data into the relevant formObj
	if (child instanceof .... ) {
      formObj.inputs.push(...)
    }
    ...

```

with this change you can just delete all of `getChildrenElements` :)

Reply:
thank you! just made the changes